compiler/luci/pass/src/FoldShapePass.cpp

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (c) 2024 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "luci/Pass/FoldShapePass.h"
+
+#include <luci/IR/CircleNodes.h>
+#include <luci/Profile/CircleNodeOrigin.h>
+
+namespace
+{
+
+template <loco::DataType OutType> luci::CircleConst *folding(luci::CircleShape *shape)
+{
+  auto input_node = loco::must_cast<luci::CircleNode *>(shape->input());
+  auto name = input_node->name();
+  auto shape_status = input_node->shape_status();
+  auto rank = input_node->rank();
+  assert(name.length() > 0);
+  if (shape_status != luci::ShapeStatus::VALID)
+    return nullptr;
+  if (rank <= 0)
```

Comment:
have you checked with rank 0 constant?

Reply:
can `rank` be minus?

Reply:
> have you checked with rank 0 constant?

The test case `rank_mismatch_NEG` in `FoldShapePass.test.cpp` is for this case!

> can `rank` be minus?

No.. AFAIK, `rank` can be a non-negative value (zero or positive integer)

Then, it might be fine just check if the `rank` is zero or not

Reply:
> The test case rank_mismatch_NEG in FoldShapePass.test.cpp is for this case!

Not this.
Please check with TF or pytorch and create a rank 0 constant and check how the frameworks act.
Please check with recipie/tflchef with rank 0 constant if this is possible or not.

Reply:
Ok, thank you for your advice!

I'll search about that, and check again! :)

Reply:
> You may have to dig to "what is rank 0 constant?"

What I found is that the constant whose rank is 0 represents scalar value

> Please check with TF or pytorch and create a rank 0 constant and check how the frameworks act.

I created a scalar value with simple tensorflow code like below, and printed the value to check its rank

```python
import tensorflow as tf

def main():
  const = tf.constant(1.0)
  rank = tf.rank(const)
  print("constant -> ", const)
  print("rank -> ", rank)
  
  return 0

if __name__ == '__main__':
  main()
```
```bash
constant ->  tf.Tensor(1.0, shape=(), dtype=float32)
rank ->  tf.Tensor(0, shape=(), dtype=int32)
```

Through this result, I concluded that `rank 0 constant` is a scalar

When I run the code a constant tensor, the result shows the rank is 1

```python
import tensorflow as tf

def main():
  const = tf.constant([1.0])
  rank = tf.rank(const)
  print("constant -> ", const)
  print("rank -> ", rank)
  
  return 0

if __name__ == '__main__':
  main()
```
```bash
constant ->  tf.Tensor([1.], shape=(1,), dtype=float32)
rank ->  tf.Tensor(1, shape=(), dtype=int32)
```

---
> Please check with recipie/tflchef with rank 0 constant if this is possible or not.

And then, I made a test model with rank 0 constant with below recipe

```bash
operand {
  name: "input1"
  type: FLOAT32
  shape {
  }
}
operand {
  name: "input2"
  type: FLOAT32
  shape {
  }
  filler {
    tag: "explicit"
    arg: "20"
  }
}
operand {
  name: "out_shape1"
  type: INT32
  shape {
  }
}
operand {
  name: "output"
  type: INT32
  shape {
  }
}
operation {
  type: "Shape"
  shape_options {
    out_type: INT32
  }
  input: "input1"
  output: "out_shape1"
}
operation {
  type: "Add"
  input: "out_shape1"
  input: "input2"
  output: "output"
}
input: "input1"
input: "input2"
output: "output"
```

![rank0](https://github.com/Samsung/ONE/assets/44961699/0441375e-1d5b-484d-93c7-c29ca54d596d)

It was possible to make a model with rank 0 constant, but it was failed to fold the `Shape` op due to the condition below (rank is 0)

```cpp
if (rank <= 0)
```

Reply:
If I change the condition to accept folding when the input is rank 0 constant like below, the result after folding `Shape` becoms `[0]` (rank 1, one-dimensional tensor with value as 0)

```cpp
if (rank < 0)
```

but I think this result is not reasonable, so it should be rejected to fold when the input is rank 0 constant
(because the scalar values have no shape, it looks strange that the output of the `Shape` has a value)

Reply:
> ... , so it should be rejected to fold when the input is rank 0 constant

How does TF or pytorch act?

Reply:
I think that you said this with same intention of the comment from https://github.com/Samsung/ONE/pull/12478#discussion_r1465561647
So I reply here together! 😃

According to the annotation from `tf.shape`, it returns an empty vector

>   `tf.shape` returns a 1-D integer tensor representing the shape of `input`.
  For a scalar input, the tensor returned has a shape of (0,) and its value is
  the empty vector (i.e. []).

So if we accept the scalar as a valid input, it seems more reasonable to return an empty vector
But I'm not sure if it is possible...